### ISSUE TITLE
Support for named parameters

### ISSUE THREAD

**Issue by hsjoberg** [dyedgreen/deno-sqlite#6]
SQLite has native support for named parameters (e.g `:var1`) instead of indexing with `?`.
My suggestion is to follow how the `mysql2` npm package does this, which gives a very pleasant experience.

From `mysql2`:
```javascript
connection.config.namedPlaceholders = true;
connection.execute('select :x + :y as z', { x: 1, y: 2 }, (err, rows) => {
});
```

Equivalent code in `deno-sqlite` could be:
```javascript
const rows = await db.query('select :x + :y as z', { x: 1, y: 2 })
```

**Comment by hsjoberg** [comment]
Suggested implementation opened here: https://github.com/dyedgreen/deno-sqlite/pull/3

**Comment by dyedgreen** [comment]
Thanks for opening the issue and submitting a PR @hsjoberg! I definitely want to support named parameters, but also want to keep the positional arguments (especially since they don't require a malloc+TextEncode to move a string into C-land).

I would suggest the following interface change:

`query` still takes variable arguments. If the first argument is an object, use the objects entries to bind values into the query; otherwise use the argument list as is done currently.

## Cases:
- 1st and only rest argument is an object -> bind object entries to query
- 1st argument is an object but there are additional arguments -> throw an error
- 1st argument is not an object -> current behaviour; bind all elements to their index

## Pros
- backwards compatible with current interface (although I don't think the module has real users yet)
- allows for both named or index based parameters, so the user has that choice
- overhead is small and constant (1 additional typeof check)

## Cons
- makes query function more complicated
  - this might be mitigated by a scheme where we do the following: Collect the object keys, build an array where each index is populated with the value from the object (indices can be retrieved from SQLite, given the key)

## Alternatives
I don't think SQLite supports introspection into what kinds or bound parameters are expected, otherwise I would have preferred to use that.

What do you think?

**Comment by hsjoberg** [comment]
@dyedgreen Thank you for your reply.

> I definitely want to support named parameters, but also want to keep the positional arguments (especially since they don't require a malloc+TextEncode to move a string into C-land).

Ah right, gotcha.

Regarding keeping the current behavior of using non-named placeholders, I agree.
However, I would suggest to overload the second argument so that it can be an object or an array, this is how for example `mysql2` does it.  
I would recommend against using the rest parameter, as it isn't very Javascripty.
In the case where the second parameter can be an array or an object, you could do a typeof check of the actual variable instead and there's no need to check the first element and the length.

This would of course break backwards compatibility, but as you noted, I seriously doubt someone is using `deno-sqlite` for a production application.

> makes query function more complicated
>    this might be mitigated by a scheme where we do the following: Collect the object keys, build an array where each index is populated with the value from the object

Fortunately it seems like `Object.entries` supports iterating over an array as well, so I suspect the code will become quite elegant.

```javascript
> Object.entries([10,20,30])
[ [ '0', 10 ], [ '1', 20 ], [ '2', 30 ] ]
``` 

## Cases:

* 2nd parameter is an object -> bind object entries to query
* 2nd parameter is an array -> bind all elements to their index
* 2nd parameter is something else -> throw an error and/or let the app crash naturally (this will probably never happen as proper use of Typescript by the end user would prevent that)

## Pros

* Small overhead and elegant checks
* Allows for both named or index based parameters
* Let's us add another parameter (e.g `options`?) in the future.

## Cons

* Breaks backwards compatibility

**Comment by dyedgreen** [comment]
@hsjoberg That sounds good to me! I think using an Array / Object as the second parameters make the most sense (especially as it allows for the API to be expanded later, if ever needed).

I can either open a new PR, or update the one you already made (albeit a lot of the changes to the docs would have to be discarded, which I think would be more work than starting from master, plus that'd avoid merge conflicts with the new error type). Let me know how you'd prefer to go about this. I'll have some time over the weekend (and possibly tomorrow morning) to get this done 😃

Thanks again for the suggestion, I'm quite happy with the new API and think this is an important feature to support 😄.

**Comment by hsjoberg** [comment]
@dyedgreen I think it makes most sense to continue on the PR/existing branch, I don't think we need much changes to be able to support arrays too. Regarding the docs, I can revert the changes made there if you want. :slightly_smiling_face: 

Cheers. :smile:

**Comment by dyedgreen** [comment]
Closed by #3
